action/selection/open_spiel/database.cc

```diff
@@ -452,6 +468,58 @@ std::string DatabaseState::ToString() const {
 
 bool DatabaseState::IsTerminal() const { return finished_; }
 
+template<class T>
+void DatabaseState::ApplyHistory(T &txn, bool use_hypopg) const {
+  // Apply all the tuning actions. This assumes that actions take zero time.
+  for (const auto &player_action : history_) {
+    // Apply the tuning action.
+    {
+      const TuningAction &action = game_->GetTuner()->GetAction(player_action.action);
+      std::string action_sql = action.sql_;
+      // If hypopg is enabled and this is a CREATE INDEX action, we need to modify the action SQL.
+      // TODO(WAN): Be more robust about cReaTE inDeX.
+      bool is_create_index = action_sql.find("CREATE INDEX") == 0 || action_sql.find("create index") == 0;
+      if (use_hypopg && is_create_index) {
+        // hypopg_create_index() is ok with the action_sql ending with a semicolon.
+        action_sql = absl::StrCat("SELECT * FROM hypopg_create_index('", action_sql, "');");
+      }
+      pqxx::result rset{txn.exec(action_sql)};
+    }
+  }
+}
+
+std::string DatabaseState::GetAppliedStateRepresentation() const {
+  // This function is used to return a string describing the current state of the database
+  // relative to the initial configuration. This function currently concatenates the
+  // actions applied in history.
+  std::stringstream actions;
+  for (const auto &player_action : history_) {
+    const TuningAction &action= game_->GetTuner()->GetAction(player_action.action);
+    actions << action.sql_ << std::endl;
+  }
+  return actions.str();
+}
+
+template<class T>
+std::pair<bool, double> DatabaseState::GetExplainAnalyzeCostUs(T &txn, const std::string &query) const {
+  std::string explain = absl::StrCat("EXPLAIN (ANALYZE, BUFFERS) ", query);
+
+  try {
+    pqxx::result rset{txn.exec(explain)};
+    std::string explain = absl::StrCat("EXPLAIN (ANALYZE, BUFFERS) ", query);
+    ExplainAnalyzeCost cost{&rset};
+    return std::make_pair(true, cost.actual_total_time_ms_ * 1000);
+  } catch (const pqxx::integrity_constraint_violation& e) {
+    // Since we're re-executing a query, it is possible that the query might abort
+    // due to failing an integrity constraint check. ExplainMicroservice does not
+    // risk aborting since it doesn't actually execute a query.
+    //
+    // TODO(wz2): Need a graceful way of identifying and possibly modeling aborts
+    std::cerr << absl::StrCat("Encountered error executing query ", explain, " ", e.what()) << std::endl;
```

Comment:
```suggestion
    std::cerr << absl::StrCat("ERROR while executing query ", explain, " ", e.what()) << std::endl;
```

either ERROR or WARN as you see fit

I suggest that we standardize on a set of prefixes (currently, only ERROR is used) to make it easy to grep and to debug from the index_picker script.